### PR TITLE
Docs: Deno and React

### PR DESCRIPTION
**Pull request by elmurci**
Add information for deno and react native in Readme.

### CODE REVIEW

**Review by intelliot — CHANGES_REQUESTED** [review]

File: README.md [hunk 1]
```diff
@@ -36,7 +36,59 @@ In an existing project (with `package.json`), install `ripple-lib`:
...
-Then see the documentation:
+Then see the [documentation](#documentation).
+### Using ripple-lib with React Native
+
+If you want to use `ripple-lib` with React Native you will need to have some of the NodeJS modules available. To help with this you can use a module like [rn-nodeify](https://github.com/tradle/rn-nodeify).
+
+1. Install dependencies
+
+    ```shell
+    npm i --save react-native-crypto
```
[the remark this answers is not in the document]
Comment (intelliot) [reply README.md:47]: `npm i --save ripple-lib`

**Comment by intelliot** [comment]
I tried for a while, but I couldn't actually get the React Native instructions to work (trying to use the current version of React Native). Maybe it only worked on a previous version? Can you try it and if it works, let me know which version you used?

File: README.md [hunk 2]
```diff
@@ -36,7 +36,59 @@ In an existing project (with `package.json`), install `ripple-lib`:
...
+    # install peer deps
+    npm i --save react-native-randombytes
+    react-native link react-native-randombytes
+    # install latest rn-nodeify
+    npm i --save-dev rn-nodeify@latest
```
[the remark this answers is not in the document]
Comment (elmurci) [reply README.md:52]: That bit is brought from the `rn-nodeify` [instructions](https://github.com/tradle/rn-nodeify)

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (elmurci) [reply README.md]: Changed to `yarn`. Altough we use `yarn` with `ripple-lib`, the user might be using `npm` with their `react-native` app though.

File: README.md [hunk 3]
```diff
@@ -36,7 +36,59 @@ In an existing project (with `package.json`), install `ripple-lib`:
...
+
+    ```shell
+    # install node core shims and recursively hack package.json files
+    # in ./node_modules to add/update the "browser"/"react-native" field with relevant mappings
+    ./node_modules/.bin/rn-nodeify --hack --install
+    ```
+
+### Using ripple-lib with Deno
+
+While official support for [Deno](https://deno.land) is added, you can use the following work-around to use `ripple-lib` with Deno:
```
[the remark this answers is not in the document]
Comment (elmurci) [reply README.md]: Done.

File: README.md [hunk 4]
```diff
@@ -36,7 +36,59 @@ In an existing project (with `package.json`), install `ripple-lib`:
...
+    // crypto is loaded first, so it can populate global.crypto
+    require('crypto')
+    ```
+
+3. After that, run the following command:
+
+    ```shell
+    # install node core shims and recursively hack package.json files
+    # in ./node_modules to add/update the "browser"/"react-native" field with relevant mappings
+    ./node_modules/.bin/rn-nodeify --hack --install
```
[the remark this answers is not in the document]
Comment (elmurci) [reply README.md]: You are correct, my bad. Done.

**Comment by elmurci** [comment]
> I tried for a while, but I couldn't actually get the React Native instructions to work (trying to use the current version of React Native). Maybe it only worked on a previous version? Can you try it and if it works, let me know which version you used?

What error did you get?

I created again a new `react-native` app and followed the steps and it worked.
Did you do `pod install` from the `ios` folder?

**Comment by mrosendin** [comment]
@intelliot I checked my `package.json` and I had the `stream` package installed:

```
    "stream": "^0.0.2",
```

I think there are additional packages that you will be required to install. I think these steps are missing from the setup instructions and should be included.

It would be ideal if `ripple-lib` was supported natively. From that perspective, adding these additional packages is a drawback of sorts. However, native support of `ripple-lib` seems like a pretty huge effort, as it would require rewriting/rewiring `crypto` and potentially other core modules. All in all what we have here works and that is good enough 👍🏻

**Comment by intelliot** [comment]
I assume this only works with React Native CLI, and does not work with Expo

**Comment by elmurci** [comment]
> @intelliot I checked my `package.json` and I had the `stream` package installed:
> 
> ```
>     "stream": "^0.0.2",
> ```
> 
> I think there are additional packages that you will be required to install. I think these steps are missing from the setup instructions and should be included.
> 
> It would be ideal if `ripple-lib` was supported natively. From that perspective, adding these additional packages is a drawback of sorts. However, native support of `ripple-lib` seems like a pretty huge effort, as it would require rewriting/rewiring `crypto` and potentially other core modules. All in all what we have here works and that is good enough 👍🏻

I don't think additional packages need to be explicitly installed.
I was able to send a prepare, sign and send a payment following the instructions above starting a project from scratch.

The one thing I have added is an extra step to explicitly state that the shim file needs to be imported.

If you think there are additional packages needed please share steps to recreate.

**Comment by intelliot** [comment]
This is the error I'm getting now:
```
 ERROR  TypeError: null is not an object (evaluating 'RNRandomBytes.seed')
```

**Comment by intelliot** [comment]
> @intelliot I checked my `package.json` and I had the `stream` package installed:
> 
> ```
>     "stream": "^0.0.2",
> ```

If I add "stream" to my package.json then I get these errors:

```
error: Error: Unable to resolve module crypto from /Users/elliot/react/AwesomeProject/node_modules/ripple-lib/dist/npm/common/hashes/sha512Half.js: crypto could not be found within the project or in these directories:
  node_modules/ripple-lib/node_modules
  node_modules

If you are sure the module exists, try these steps:
 1. Clear watchman watches: watchman watch-del-all
 2. Delete node_modules and run yarn install
 3. Reset Metro's cache: yarn start --reset-cache
 4. Remove the cache: rm -rf /tmp/metro-*
  1 | "use strict";
  2 | Object.defineProperty(exports, "__esModule", { value: true });
> 3 | const crypto_1 = require("crypto");
    |                           ^
  4 | const sha512Half = (hex) => {
  5 |     return crypto_1.createHash('sha512')
  6 |         .update(Buffer.from(hex, 'hex'))
    at ModuleResolver.resolveDependency (/Users/elliot/react/AwesomeProject/node_modules/metro/src/node-haste/DependencyGraph/ModuleResolution.js:234:15)
    at DependencyGraph.resolveDependency (/Users/elliot/react/AwesomeProject/node_modules/metro/src/node-haste/DependencyGraph.js:413:43)
    at Object.resolve (/Users/elliot/react/AwesomeProject/node_modules/metro/src/lib/transformHelpers.js:317:42)
    at resolve (/Users/elliot/react/AwesomeProject/node_modules/metro/src/DeltaBundler/traverseDependencies.js:629:33)
    at /Users/elliot/react/AwesomeProject/node_modules/metro/src/DeltaBundler/traverseDependencies.js:645:26
    at Array.reduce (<anonymous>)
    at resolveDependencies (/Users/elliot/react/AwesomeProject/node_modules/metro/src/DeltaBundler/traverseDependencies.js:644:33)
    at /Users/elliot/react/AwesomeProject/node_modules/metro/src/DeltaBundler/traverseDependencies.js:329:33
    at Generator.next (<anonymous>)
    at asyncGeneratorStep (/Users/elliot/react/AwesomeProject/node_modules/metro/src/DeltaBundler/traverseDependencies.js:137:24)
 BUNDLE  ./index.js 

error: Error: Unable to resolve module crypto from /Users/elliot/react/AwesomeProject/node_modules/ripple-lib/dist/npm/common/hashes/sha512Half.js: crypto could not be found within the project or in these directories:
  node_modules/ripple-lib/node_modules
  node_modules

If you are sure the module exists, try these steps:
 1. Clear watchman watches: watchman watch-del-all
 2. Delete node_modules and run yarn install
 3. Reset Metro's cache: yarn start --reset-cache
 4. Remove the cache: rm -rf /tmp/metro-*
  1 | "use strict";
  2 | Object.defineProperty(exports, "__esModule", { value: true });
> 3 | const crypto_1 = require("crypto");
    |                           ^
  4 | const sha512Half = (hex) => {
  5 |     return crypto_1.createHash('sha512')
  6 |         .update(Buffer.from(hex, 'hex'))
    at ModuleResolver.resolveDependency (/Users/elliot/react/AwesomeProject/node_modules/metro/src/node-haste/DependencyGraph/ModuleResolution.js:234:15)
    at DependencyGraph.resolveDependency (/Users/elliot/react/AwesomeProject/node_modules/metro/src/node-haste/DependencyGraph.js:413:43)
    at Object.resolve (/Users/elliot/react/AwesomeProject/node_modules/metro/src/lib/transformHelpers.js:317:42)
    at resolve (/Users/elliot/react/AwesomeProject/node_modules/metro/src/DeltaBundler/traverseDependencies.js:629:33)
    at /Users/elliot/react/AwesomeProject/node_modules/metro/src/DeltaBundler/traverseDependencies.js:645:26
    at Array.reduce (<anonymous>)
    at resolveDependencies (/Users/elliot/react/AwesomeProject/node_modules/metro/src/DeltaBundler/traverseDependencies.js:644:33)
    at /Users/elliot/react/AwesomeProject/node_modules/metro/src/DeltaBundler/traverseDependencies.js:329:33
    at Generator.next (<anonymous>)
    at asyncGeneratorStep (/Users/elliot/react/AwesomeProject/node_modules/metro/src/DeltaBundler/traverseDependencies.js:137:24)
 BUNDLE  ./index.js 

error: Error: Unable to resolve module crypto from /Users/elliot/react/AwesomeProject/node_modules/ripple-lib/dist/npm/common/hashes/sha512Half.js: crypto could not be found within the project or in these directories:
  node_modules/ripple-lib/node_modules
  node_modules

If you are sure the module exists, try these steps:
 1. Clear watchman watches: watchman watch-del-all
 2. Delete node_modules and run yarn install
 3. Reset Metro's cache: yarn start --reset-cache
 4. Remove the cache: rm -rf /tmp/metro-*
  1 | "use strict";
  2 | Object.defineProperty(exports, "__esModule", { value: true });
> 3 | const crypto_1 = require("crypto");
    |                           ^
  4 | const sha512Half = (hex) => {
  5 |     return crypto_1.createHash('sha512')
  6 |         .update(Buffer.from(hex, 'hex'))
    at ModuleResolver.resolveDependency (/Users/elliot/react/AwesomeProject/node_modules/metro/src/node-haste/DependencyGraph/ModuleResolution.js:234:15)
    at DependencyGraph.resolveDependency (/Users/elliot/react/AwesomeProject/node_modules/metro/src/node-haste/DependencyGraph.js:413:43)
    at Object.resolve (/Users/elliot/react/AwesomeProject/node_modules/metro/src/lib/transformHelpers.js:317:42)
    at resolve (/Users/elliot/react/AwesomeProject/node_modules/metro/src/DeltaBundler/traverseDependencies.js:629:33)
    at /Users/elliot/react/AwesomeProject/node_modules/metro/src/DeltaBundler/traverseDependencies.js:645:26
    at Array.reduce (<anonymous>)
    at resolveDependencies (/Users/elliot/react/AwesomeProject/node_modules/metro/src/DeltaBundler/traverseDependencies.js:644:33)
    at /Users/elliot/react/AwesomeProject/node_modules/metro/src/DeltaBundler/traverseDependencies.js:329:33
    at Generator.next (<anonymous>)
    at asyncGeneratorStep (/Users/elliot/react/AwesomeProject/node_modules/metro/src/DeltaBundler/traverseDependencies.js:137:24)
```

**Comment by mrosendin** [comment]
@intelliot you probably need to run the `rn-nodeify --hack` command again. I basically run it every time I add a new package.

**Comment by intelliot** [comment]
@mrosendin thanks for the reply, but even after running that, I'm still getting the error:
```
 ERROR  TypeError: null is not an object (evaluating 'RNRandomBytes.seed')
 ERROR  Invariant Violation: Module AppRegistry is not a registered callable module (calling runApplication). A frequent cause of the error is that the application entry file path is incorrect. 
      This can also happen when the JS bundle is corrupt or there is an early initialization error when loading React Native.
 ERROR  Invariant Violation: Module AppRegistry is not a registered callable module (calling runApplication). A frequent cause of the error is that the application entry file path is incorrect. 
      This can also happen when the JS bundle is corrupt or there is an early initialization error when loading React Native.
```

If you happen to have time, do you see any issue with my project? Maybe it's a problem with my dev environment... https://github.com/intelliot/react-native-ripple-lib

**Comment by mrosendin** [comment]
The linked project issues were resolved by modifying the Podfile. Here's a summary of what I did:

> Steps taken with the example proj with Node v14.15.3

- `yarn`
- `cd ios && rm Podfile.lock && pod install`
- run `./node_modules/rn-nodeify --hack --install`
- `npx react-native run-ios` (I'm using `npx` here instead of the RN CLI)
- saw `duplicate symbols for architecture x86_64` in the build errors. problem seems to be with `react-native-udp` being doubly linked. after a bit of debugging I saw that Flipper was enabled in `Podfile`. I commented it out:

```
  # Uncomment the code below to enable Flipper.
  #
  # You should not install Flipper in CI environments when creating release
  # builds, this will lead to significantly slower build times.
  #
  # Note that if you have use_frameworks! enabled, Flipper will not work.
  #
  #  use_flipper!
  #  post_install do |installer|
  #    flipper_post_install(installer)
  #  end
```

- run `cd ios && rm Podfile.lock && pod install` again
- `npx react-native run-ios`

Got an error this time about TCP sockets:

```
❌  duplicate symbol '_GCDAsyncSocketUseCFStreamForTLS' in
> libCocoaAsyncSocket.a(GCDAsyncSocket.o)
> libTcpSockets.a(GCDAsyncSocket.o)
```

**Podfile**

```ruby
require_relative '../node_modules/react-native/scripts/react_native_pods'
require_relative '../node_modules/@react-native-community/cli-platform-ios/native_modules'

platform :ios, '10.0'

target 'AwesomeProject' do
  config = use_native_modules!

  use_react_native!(
    :path => config[:reactNativePath],
    # to enable hermes on iOS, change `false` to `true` and then install pods
    :hermes_enabled => false
  )

  target 'AwesomeProjectTests' do
    inherit! :complete
    # Pods for testing
  end

  # Enables Flipper.
  #
  # Note that if you have use_frameworks! enabled, Flipper will not work and
  # you should disable the next line.
  # use_flipper!()

  # post_install do |installer|
  #   react_native_post_install(installer)
  # end
end
```

I just saw https://github.com/Rapsssito/react-native-tcp-socket/issues/36 and https://github.com/Rapsssito/react-native-tcp-socket/issues/61 which lead me to believe that it's best solved from within xCode itself. I will try to investigate further tomorrow.

**Comment by mrosendin** [comment]
Nvm, I thought I could just try in xCode really quickly and the build succeeded! Follow the solution [here](https://github.com/Rapsssito/react-native-tcp-socket/issues/61) but replace `CocoaAsyncSocket ` with `GCDAsyncSocket`.

**Comment by elmurci** [comment]
> @mrosendin thanks for the reply, but even after running that, I'm still getting the error:
> 
> ```
>  ERROR  TypeError: null is not an object (evaluating 'RNRandomBytes.seed')
>  ERROR  Invariant Violation: Module AppRegistry is not a registered callable module (calling runApplication). A frequent cause of the error is that the application entry file path is incorrect. 
>       This can also happen when the JS bundle is corrupt or there is an early initialization error when loading React Native.
>  ERROR  Invariant Violation: Module AppRegistry is not a registered callable module (calling runApplication). A frequent cause of the error is that the application entry file path is incorrect. 
>       This can also happen when the JS bundle is corrupt or there is an early initialization error when loading React Native.
> ```
> 
> If you happen to have time, do you see any issue with my project? Maybe it's a problem with my dev environment... https://github.com/intelliot/react-native-ripple-lib

@intelliot I was able to run your project with no issues (after fixing the `GCDAsyncSocket` problem as @mrosendin pointed out)

**Review by intelliot — APPROVED** [review]

**Pull request merged**